Feature-Backup codes support for authentication flow

Mahir18-GIT opened the pull request:
Summary

This PR introduces backup code support for authentication. Users can now generate a set of one-time use backup codes, verify them during login, and regenerate new codes if needed.

I have tested the complete authentication flow in Postman (with screenshots attached) to confirm the feature works end-to-end.

Changes

Controllers:

- Added backupCode.controller.js with endpoints to generate, verify, and regenerate codes.

Models:

- Updated user.model.js to include a backupCodes field with metadata (hash, used, createdAt, usedAt).

Utils:

- Created backup.util.js to handle secure code generation, hashing, formatting, and validation.

Routes:

- Extended auth.routes.js to register three new routes:

POST /generate-backup-codes

POST /verify-backup-code

POST /regenerate-backup-codes

Notes

Codes are only returned in plaintext once (at generation or regeneration).

bcrypt hashing is used to securely store codes in the database.

Input normalization added to avoid issues with spaces/dashes.

Error handling and generic failure messages are in place to avoid leaking details.

Testing Performed:

 JWT Login Flow – verified login returns a valid token.
 Generate Backup Codes – successfully generated and stored hashed codes.
 Verify with Backup Code – tested successful login with backup code (token issued).
 Regenerate Codes – tested regeneration both with and without JWT:

With JWT → allowed regeneration.

Without JWT → denied as expected.
 Postman Evidence – screenshots attached showing each stage (login, generate, verify, regenerate).

<img width="1919" height="1035" alt="image" src="https://github.com/user-attachments/assets/161ede07-63ed-435c-a574-9f11de36ef87" />
<img width="1919" height="999" alt="image" src="https://github.com/user-attachments/assets/e050cc0e-fa7c-421c-8da2-d09b2b0c0e90" />
<img width="1919" height="1034" alt="image" src="https://github.com/user-attachments/assets/02206b91-8ae2-4011-bf96-2b29028385a3" />
<img width="1919" height="1034" alt="image" src="https://github.com/user-attachments/assets/17298cfb-9cf3-4f34-9fb3-5fa7b6baf23f" />

- Generated backup codes with JWT Authenticatio:
<img width="1918" height="1030" alt="image" src="https://github.com/user-attachments/assets/25ef8a94-41c2-4b24-90c4-f932f6b1bb26" />

- Without Authentication, it gives an error:
<img width="1919" height="1031" alt="image" src="https://github.com/user-attachments/assets/6941b831-95b9-4a0b-a371-1fc46245e846" />

Logged in with one of these codes-
<img width="1919" height="1034" alt="image" src="https://github.com/user-attachments/assets/245f81d4-51a1-4de5-ac70-f57cab211296" />

Regenerated Backup Codes with JWT Authentication:
<img width="1919" height="1036" alt="image" src="https://github.com/user-attachments/assets/a5702462-cfb2-492f-b9d8-0948b24a2f77" />

And without the token, it does not generate:
<img width="1919" height="1032" alt="image" src="https://github.com/user-attachments/assets/e091e3ec-b925-4fe7-8cb9-bb6a04c15d84" />

<img width="1919" height="1000" alt="image" src="https://github.com/user-attachments/assets/2b931859-44d0-4871-b1bd-9bcc5265dbbd" />

swagboyswastik committed:
feat(auth): rebasing dev into the featured branch

swagboyswastik committed:
chore: update lockfile after npm install

Mahir18-GIT committed:
chore(deps): regenerate package-lock.json after merge/rebase

Ed-ward-Harris commented:
The email service file adds new ENV variables, I like this, but you haven't added them to the example env file.
You need to add new ENV variable to example.env, pre fill with:
EMAIL_HOST=smtp.gmail.com
EMAIL_PORT=465
Also in the email service file secure should be set to true  to match gmail preferences. 

Same with the backup code utils file. You have added new environment variables, so the example env file should be updated accordingly. 

All the endpoints work as expect., which is great. The issue is that the regenerate and generate APIs require zero authentication. This means all you need to request new codes is an email address. This is highly insecure and needs to be resolved.

Mahir18-GIT committed:
docs(env): update .env.example with email + backup-code vars (Mailtrap option)

Mahir18-GIT committed:
fix(auth): protect backup-code generate/regenerate with JWT (authMiddleware)

Mahir18-GIT commented:
Thanks a lot for the detailed review, Edward.

I’ve made the following changes based on your feedback:

Updated .env.example to include the new email variables (EMAIL_HOST, EMAIL_PORT, EMAIL_SECURE) and backup code settings.

Set the default values to align with Mailtrap preferences, with clear placeholders for user-specific credentials.

Updated the auth routes so that generate and regenerate backup code endpoints are now protected with authMiddleware.

Let me know if this addresses everything or if there’s anything else you’d like adjusted

Proof that its working with JWT authentication(both generate and regenerate) and gives error without Authentication:
<img width="1919" height="1013" alt="Screenshot 2025-09-23 133941" src="https://github.com/user-attachments/assets/9f56f927-3e1c-4123-9aa1-53af5f12e22f" />

Without Authetication:
<img width="1919" height="1031" alt="Screenshot 2025-09-23 134106" src="https://github.com/user-attachments/assets/4291cb1d-2a00-4282-a847-57ef8011c8e8" />

Same for Regenerate Backup Codes:
<img width="1919" height="1036" alt="Screenshot 2025-09-23 134340" src="https://github.com/user-attachments/assets/e30df2d4-2747-4cdf-aae1-d787b5de8285" />

Without it:
<img width="1919" height="1032" alt="Screenshot 2025-09-23 134424" src="https://github.com/user-attachments/assets/c04153c9-3580-40ed-a086-80b7f354e73a" />

Mahir18-GIT committed:
chore(deps): add zod  and ua-parser-js (fix module-not-found)

Mahir18-GIT committed:
feat(email): respect EMAIL_SECURE and default example to Gmail (465/TLS)
